tsl/src/nodes/skip_scan/planner.c

```diff
@@ -0,0 +1,599 @@
+	double total = index_path->path.total_cost;
+	double rows = index_path->path.rows;
+
+	if (index_path->indexinfo->sortopfamily == NULL)
+		return NULL; /* non-orderable index, skip these for now */
+
+	/* orderbyops are not compatible with skipscan */
+	if (index_path->indexorderbys != NIL)
+		return NULL;
+
+	SkipScanPath *skip_scan_path = (SkipScanPath *) newNode(sizeof(SkipScanPath), T_CustomPath);
+
+	skip_scan_path->cpath.path.pathtype = T_CustomScan;
+	skip_scan_path->cpath.path.pathkeys = index_path->path.pathkeys;
+	skip_scan_path->cpath.path.pathtarget = index_path->path.pathtarget;
+	skip_scan_path->cpath.path.param_info = index_path->path.param_info;
+	skip_scan_path->cpath.path.parent = index_path->path.parent;
+	skip_scan_path->cpath.path.rows = ndistinct;
+	skip_scan_path->cpath.custom_paths = list_make1(index_path);
+	skip_scan_path->cpath.methods = &skip_scan_path_methods;
+
+	/* We calculate SkipScan cost as ndistinct * startup_cost + (ndistinct/rows) * total_cost
+	 * ndistinct * startup_cost is to account for the rescans we have to do and since startup
+	 * cost for indexes does not include page access cost we add a fraction of the total cost
+	 * accounting for the number of rows we expect to fetch.
+	 */
+	skip_scan_path->cpath.path.startup_cost = index_path->path.startup_cost;
+	skip_scan_path->cpath.path.total_cost = ndistinct * startup + (ndistinct / rows) * total;
+
+	/* While add_path may pfree paths with higher costs
+	 * it will never free IndexPaths and only ever do a shallow
+	 * free so reusing the IndexPath here is safe. */
+	skip_scan_path->index_path = index_path;
+
+	/* find the ordering operator we'll use to skip around each key column
+	 * Since the DISTINCT columns are required to be prefix of ORDER BY
+	 * and we only support single column DISTINCT the first pathkey
+	 * has to be the distinct column
+	 */
+	PathKey *first_pathkey = linitial(index_path->path.pathkeys);
+	TargetEntry *tle = get_tle_for_pathkey(index_path->indexinfo->indextlist, first_pathkey);
+
+	/* SkipScan on expressions not supported */
+	if (!IsA(tle->expr, Var))
+		return NULL;
+
+	/* build skip qual this may fail if we cannot look up the operator */
+	if (!build_skip_qual(skip_scan_path, index_path, castNode(Var, tle->expr)))
+		return NULL;
+
+	return skip_scan_path;
+}
+
+/*
+ * Creates SkipScanPath for each path of subpaths that is an IndexPath
+ * If no subpath can be changed to SkipScanPath returns NULL
+ * otherwise returns list of new paths
+ */
+static List *
+build_subpath(PlannerInfo *root, List *subpaths, double ndistinct)
+{
+	bool has_skip_path = false;
+	List *new_paths = NIL;
+	ListCell *lc;
+
+	foreach (lc, subpaths)
+	{
+		Path *child = lfirst(lc);
+		if (IsA(child, IndexPath))
+		{
+			SkipScanPath *skip_path =
+				skip_scan_path_create(root, castNode(IndexPath, child), ndistinct);
+
+			if (skip_path)
+			{
+				child = (Path *) skip_path;
+				has_skip_path = true;
+			}
+		}
```

Comment:
Why does this only check for `IndexPath` while the "upper" function has checks for `IndexOnlyPath` as well?

Reply:
there is no IndexOnlyPath

Reply:
Ah, mixed them up. Was thinking of `IndexOnlyScan`, which is not relevant here.